Topics/System_Design/Handling_100K_Requests.md

```diff
@@ -0,0 +1,52 @@
+
+## Load Balancing
+![Load Balancing](./assets/load-balancer.png)
+*Image from NetScaler*
+
+A **load balancer** is responsible for routing requests from our clients to an instance of our APIs. This will enable the horizontal scaling we discussed previously. Some load balancing strategies include distributing requests across instances sequentially or routing requests to the instances with the least amount of traffic. A load balancer also improves reliability of our system, if one of our servers go down, we can reroute traffic to another server instead of having the entire system fail. Additionally, we can set up a load balancer to increase the number of instances when the number of requests exceeds the capacity of the current instances or to reduce the number of instances when we might expect less traffic. 
+
+
+## Database Scaling
+![Primary-Replica Database](./assets/primary-replica.png)
+*Image from Apurva Agrawal's Medium*
+
+We will focus on scaling relational databases in this section. There exist many techniques such as **replication**, **federation**, **sharding** and so on. We will briefly touch on **replication** and defer to the sources below for more detailed information on the other techniques. **Replication** refers to the concept of replicating data from a primary database to replica databases such that any updates to the database requested by the client are first written to the primary before then being propgated to each replica. Then, the other clients and servers can read from any of the replicas to distribute the workload from all our requests.  
+
+
+## Caching
+![Redis Cache](./assets/redis-cache.png)
+*Image from Backendless*
+
+Another way we can reduce the load on our databases is by introducing a cache layer using something like [Redis](https://redis.io/). When a user makes a request for some data from the database, we store that information in the cache layer with some expiration time *t*. Any future requests made for that information within time period *t* are actually hitting the cache rather than the database. This reduces the workload of our database by handling some of it with the cache layer, and we can scale our cache layer as well by using distributed caching and storing data across multiple machines. 
+
+
+
+## Content Delivery Networks (CDNs) 
+CDN servers keep a cache of our files (images, HTML/CSS, JS, ...) at data centers close in geographical proximity to users across the world. When user's make a request for data, it is retrieved from the server closest to the user. This again offloads some of the requests from the main server and helps with reducing errors by increasing redundancy. 
```

Comment:
Since you have used picture to illustrate ideas for all previous sections, I think you probably want to use one for CDN as well. There are a wide range of uses of CDNs, from picture/manga websites to frameworks like bootstrap.
![image](https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/93541585/c1f2ce9d-fdd0-42f1-a44e-21b4d007aaf2)

![image](https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/93541585/c321acf8-b86a-43ec-bfc9-25c09026d1c4)

Reply:
thanks I included another image as you suggested